java/apirag-starter/src/main/java/com/datasqrl/ai/LowLevelAgent.java

```diff
@@ -0,0 +1,155 @@
+          .model(chatConfig.getModelName())
+          .messages(contextWindow.getMessages())
+          .functions(contextWindow.getFunctions())
+          .functionCall("auto")
+          .n(1)
+          .topP(chatConfig.getTopP())
+          .temperature(chatConfig.getTemperature())
+          .maxTokens(chatConfig.getMaxOutputTokens())
+          .logitBias(new HashMap<>())
+          .build();
+      Flowable<ChatCompletionChunk> flowable = service.streamChatCompletion(chatCompletionRequest);
+
+      AtomicBoolean isFirst = new AtomicBoolean(true);
+      AssistantMessage responseMessage = service.mapStreamToAccumulator(flowable)
+          .doOnNext(accumulator -> {
+            if (accumulator.isFunctionCall()) {
+              if (isFirst.getAndSet(false)) {
+                System.out.println("Executing function " + accumulator.getAccumulatedChatFunctionCall().getName() + "...");
+              }
+            } else {
+              if (isFirst.getAndSet(false)) {
+                System.out.print("Response: ");
+              }
+              if (accumulator.getMessageChunk().getContent() != null) {
+                System.out.print(accumulator.getMessageChunk().getContent());
+              }
+            }
+          })
+          .doOnComplete(System.out::println)
+          .lastElement()
+          .blockingGet()
+          .getAccumulatedMessage();
+      session.addMessage(responseMessage);
+
+      ChatFunctionCall functionCall = responseMessage.getFunctionCall();
+      if (functionCall != null) {
+        FunctionValidation<ChatMessage> functionValidation = session.validateFunctionCall(functionCall);
+        if (functionValidation.isValid()) {
+          log.info("Executing {} with arguments {}", functionCall.getName(), functionCall.getArguments().toPrettyString());
+          ChatMessage functionResponse = session.executeFunctionCall(functionCall, context);
+          log.info("Executed {} with results: {}", functionCall.getName(), functionResponse.getTextContent());
+          session.addMessage(functionResponse);
+        } else {
+          log.info("Function call {} failed.", functionCall.getName());
+          session.addMessage(new FunctionMessage("{\"error\": \"" + functionValidation.validationError().errorMessage() + "\"}", "error"));
+        }
+      } else {
+        System.out.print("Next Query: ");
+        String nextLine = scanner.nextLine();
+        if (nextLine.equalsIgnoreCase("exit")) {
+          System.exit(0);
+        }
+        ChatMessage nextMsg = new UserMessage(nextLine);
+        session.addMessage(nextMsg);
+      }
+    }
+  }
+
+  public static void main(String... args) throws Exception {
+    String systemPrompt = "You are an incredibly enthusiastic and joyous life coach. You provide advice to people. You look up any and all advice you give via the provided functions.";
+    Path toolsPath = Path.of("java", "apirag-starter", "src", "main", "resources", "tools", "advice.tools.json");
```

Comment:
This path looks strange to me. If you run this in Intellij it should be just relative to the module path, no?

Reply:
I tried, but if I run the java file by itself, it starts in the home folder. 🤷

Reply:
Got it. How are you running it?

Reply:
![Screenshot 2024-07-17 at 22 05 39](https://github.com/user-attachments/assets/b7756db1-5464-47a9-93a6-615bb6a54814)
LIke this